feat: Add async functionality to providers

leohoare opened the pull request:
## This PR
Adds the ability for open feature providers to use async methods
It extends the single client and attempts to refactor some code

Related Issues
https://github.com/open-feature/python-sdk/issues/284
https://github.com/open-feature/python-sdk/pull/383
https://github.com/open-feature/python-sdk/pull/385

### Follow-up Tasks & TODOS

- [x] Add tests for async provider and implementation
- [x] Add documentation and how to implement an async hook

beeme1mr commented:
Hey @leohoare, this looks good so far. Could you please add tests covering async providers with sync client calls and vise versa? Thanks for your hard work on this. 🍻

leohoare commented:
Thanks @beeme1mr, I've added some tests and addressed the coverage issues.

One thing to note is sync methods are always enforced on async providers.
you can't implement an async only provider, although, you can work around this unless using `NotImplementedError` on the sync methods. It was implemented on the AbstractProvider to keep current functionality the same.

Is this clear enough from the documentation?

beeme1mr commented:
> Thanks @beeme1mr, I've added some tests and addressed the coverage issues.
> 
> One thing to note is sync methods are always enforced on async providers.
> you can't implement an async only provider, although, you can work around this unless using `NotImplementedError` on the sync methods. It was implemented on the AbstractProvider to keep current functionality the same.
> 
> Is this clear enough from the documentation?

Sorry, I didn't get a chance to look at this today. It's on my to-do for tomorrow.

leohoare commented:
No rush, I'll be off grid over the weekend anyway.

beeme1mr reviewed (commented):
Hey @leohoare, this looks good from what I can tell but I wouldn't consider myself a Python expert.

I see you have tests but would you mind also enumerating the expected behavior for the following scenarios?

- performing an async evaluation on a synonymous provider
- performing a sync evaluation on an async provider that implements the AbstractProvider 

I believe I understand how everything will behave but I'd like to confirm.

Also, could someone with more Python experience please weigh in when you have a moment? FYI, @aepfli @guidobrei @federicobond @jamescarr @lukas-reining @toddbaert

leohoare commented:
> I see you have tests but would you mind also enumerating the expected behavior for the following scenarios?

Do you mean explain the scenarios or update the tests?

> performing an async evaluation on a synonymous provider

If async evaluation is not implemented, it will fall back to calling the synchronous function. 
I.e. the call will be async but the code itself will be blocking and not truly async.

> performing a sync evaluation on an async provider that implements the AbstractProvider

a provider that implements async calls is forced to implement sync functions.
If calls on the AbstractProvider are implemented in the sync function, then the calls will be like client previously functioned.
```
    def resolve_boolean_details(
        self,
        flag_key: str,
        default_value: bool,
        evaluation_context: Optional[EvaluationContext] = None,
    ) -> FlagResolutionDetails[bool]:
        # do some thing things...
        return FlagResolutionDetails(value=True)
```
If the provider chooses to only implement async functions and throw an error on the sync functions.
Then the client will throw an error when attempting to resolve values.
```
        def resolve_boolean_details(
            self,
            flag_key: str,
            default_value: bool,
            evaluation_context: Optional[EvaluationContext] = None,
        ) -> FlagResolutionDetails[bool]:
            raise NotImplementedError("Use the async method")
```
We're essentially offloading the decision to the provider on how to handle async/sync calls. Implementing the async calls is optional and defaults to sync when not implemented.

leohoare commented:
Sorry keep forgetting to sign-off the commits -.-

beeme1mr commented:
I've approved because I'm good with the approach. Hopefully others with more Python experience can also provide some thoughts.

lukas-reining commented:
Sorry for the late reply, I have been on vacation. Will have a look in the next 1 or 2 days :)

ChihweiLHBird reviewed (approved):
This looks good to me! I think we should mention the fallback mechanism in the documentation to avoid confusion. Users might expect asynchronous execution when calling those async functions, but they actually get synchronous function execution.

leohoare commented:
I've updated the readme to include the suggestion @ChihweiLHBird, as well a general usage code block.

@chrfwow due to the refactoring required in this PR, some of your recent merge had to be moved. 
Could you please also review this PR (particularly the last two commits)?

beeme1mr commented on README.md:
````diff
@@ -316,6 +317,26 @@ async def some_endpoint():
+```
+
+See the [develop a provider](#develop-a-provider) for how to support asynchronous functionality in providers.
+
````
```suggestion
```

Please remove the extra carriage return.

beeme1mr commented on README.md:
````diff
@@ -390,6 +411,57 @@ class MyProvider(AbstractProvider):
+
+```
+
+
````
```suggestion
```

Please remove the extra carriage return. This will be flagged by the linter when it's included in the docs page. Thanks!

chrfwow reviewed (changes requested):
The changes of my latest additions seem fine so far, but I have a few other questions

leohoare commented on openfeature/client.py:
```diff
@@ -295,54 +461,224 @@ def evaluate_flag_details(  # noqa: PLR0915
 
+        return provider, hook_context, hook_hints, merged_hooks, reversed_merged_hooks
+
+    def _assert_provider_status(
```
How about `_validate_provider_status`? or do you have any other suggestions?

chrfwow replied:
Wouldn't it be possible to just throw the proper exception (both `ProviderNotReadyError` and `ProviderFatalError` are instances of `OpenFeatureError`, so the exception handling should work)? This function is called inside the try catch block, so the error hooks will also be called this way.
If not, I would go for a name like `_handle_provider_not_ready`, but I don't like that name either

leohoare replied:
This code previously wasn't in the try catch. Although with your refactor, we should be able to throw exceptions here now.
I'll look into this today.

leohoare replied:
Good catch, this simplified the logic a lot.

leohoare commented on openfeature/client.py:
```diff
@@ -402,6 +738,48 @@ def evaluate_flag_details(  # noqa: PLR0915
+        default_value: typing.Any,
+        evaluation_context: typing.Optional[EvaluationContext] = None,
+    ) -> FlagEvaluationDetails[typing.Any]:
+        args = (
```
The way it's used is just syntactic sugar to get around strict typing.
We call it like so:
```
args = ...
resolution = await get_details_callable(*args)
```

The alternative is passing in the arguments by key.
This option likely has a typing error as we're dynamically resolving the function call.
```
resolution = await get_details_callable(flag_key=flag_key,...)
```

Another option is arguments by order.
With this option, we lose the ability to have multiple defaults (unless you pass in arguments in the exact order) and order must match exactly, so it's not often recommended.
``` 
resolution = await get_details_callable(flag_key, default_value, evaluation_context)
```

When you pass it in like *args, it's a way of dynamically doing option 1. 
This pattern was also copied from the sync function.

gruebel replied:
this is incorrect. passing `*args` to `get_details_callable()` unpacks them as positional arguments and therefore you can directly set them there. you are mistaken it with `**kwargs`. Feel free to try it out and switch the position in the `args` tuple and you will see how it messes up the function call.

Nevertheless I need to check, if it is possible to properly type the dynamic resolving, but not important in this PR, because it is done the same way in the sync variant as you mentioned.

leohoare replied:
Ah true, misinterpreted this sorry.
Happy to update this to `**kwargs` or if you want to avoid scope creep, I can do this in a follow up PR.

gruebel replied:
Nono, leave as is and feel free to give it a try in a separate PR, because the type checker won't be so nice to you

leohoare commented on openfeature/client.py:
```diff
@@ -402,6 +738,48 @@ def evaluate_flag_details(  # noqa: PLR0915
                 hook_hints,
             )
 
+    async def _create_provider_evaluation_async(
```
Yeah, they need to be separate functions as this is needs to propagate the async function.
One big downside of python's handling of async is it often leads to forked/duplicated code.

One option is to convert the logic before/after the _get_details_callable for both functions into helper functions, however, they're already only a few calls to helpers. 

I'll see what I can do to clean this up

chrfwow replied:
I see... Then let's leave it as is 👍

leohoare commented on openfeature/client.py:
```diff
@@ -402,6 +738,48 @@ def evaluate_flag_details(  # noqa: PLR0915
+            default_value,
+            evaluation_context,
+        )
+        get_details_callables_async: typing.Mapping[
```
Yep, followed the pattern from the sync call but it could be global.

gruebel replied:
this is a perfect use case for structural pattern matching, which we can use in 1-2 years, when the minimum supported Python version reaches 3.10 😅 but till then this is good enough.

leohoare commented on openfeature/client.py:
```diff
@@ -295,54 +461,224 @@ def evaluate_flag_details(  # noqa: PLR0915
+                hook_hints,
+            )
+
+    def evaluate_flag_details(
```
> Does the async prevent the extraction into a common function, or is there another reason for the duplication?

Somewhat yes, because the async call is in the middle of the try catch if we wanted to split it into common code we could only do something like:
* calls at the start of try in common call
```
            error_code = self._assert_provider_status(
                flag_type,
                hook_context,
                reversed_merged_hooks,
                hook_hints,
            )
            if error_code:
                flag_evaluation = FlagEvaluationDetails(
                    flag_key=flag_key,
                    value=default_value,
                    reason=Reason.ERROR,
                    error_code=error_code,
                )
                return flag_evaluation

            merged_context = self._before_hooks_and_merge_context(
                flag_type,
                hook_context,
                merged_hooks,
                hook_hints,
                evaluation_context,
            )
```
* calls after the async call in common code
```
            after_hooks(
                flag_type,
                hook_context,
                flag_evaluation,
                reversed_merged_hooks,
                hook_hints,
            )

            return flag_evaluation
```
* A helper call for each of the exception

I didn't really deem this worth it and would lead to passing around a lot of variables to common functions.
What are your thoughts?

chrfwow replied:
Let's leave it as is 👍

gruebel reviewed (commented):
nice work, just added a few comments

leohoare commented on openfeature/provider/in_memory_provider.py:
```diff
@@ -117,3 +157,10 @@ def _resolve(
             raise FlagNotFoundError(f"Flag '{flag_key}' not found")
         return flag.resolve(evaluation_context)
+
+    async def _resolve_async(
+        self,
+        flag_key: str,
+        evaluation_context: typing.Optional[EvaluationContext],
+    ) -> FlagResolutionDetails[V]:
+        return self._resolve(flag_key, evaluation_context)
```
It was just to demonstrate calling an async function, I can update it to directly calling _resolve if we prefer.

gruebel replied:
Thought about it again, let's leave it. The `InMemoryProvider` is always a nice and easy example on how to create your own provider

beeme1mr commented:
Greets everyone! I'd like to get this merged by the end of the week if possible. Please leave you feedback ASAP if you have any concerns. Thanks!

@leohoare, thanks for you hard work and patience. It's important that we have consensus when making changes to the public APIs. So changes like this tend to take a while.

lukas-reining commented on README.md:
```diff
@@ -109,6 +109,7 @@ print("Value: " + str(flag_value))
 | ✅      | [Eventing](#eventing)                                               | React to state changes in the provider or flag management system.                                                                     |
 | ✅      | [Shutdown](#shutdown)                                               | Gracefully clean up a provider during application shutdown.                                                                           |
 | ✅      | [Transaction Context Propagation](#transaction-context-propagation) | Set a specific [evaluation context](/docs/reference/concepts/evaluation-context) for a transaction (e.g. an HTTP request or a thread) |
+| ✅      | [Asynchronous Feature Retrieval](#asynchronous-feature-retrieval)   | Evaluate flags in an asychronous context.                                                                                                 |
```
Not sure if we should add this here. 
For JS we do not consider this as a feature and the feature list generally lists features that are "explicitly specified".
So for consistency I think it would make sense to leave it out or alternatively add it to the JS SDK readme.

leohoare replied:
@lukas-reining fair call, I'll leave it out.
Just the row in the table or the section `Asynchronous Feature Retrieval` as well?

lukas-reining replied:
I would say only in the table.
Having it in the README can make sense to me :)

lukas-reining replied:
I would say only in the table.
Having it in the README can make sense to me :)

leohoare replied:
Alright, I've removed the row from the table.

beeme1mr commented:
I'll merge this PR tomorrow unless someone objects.
